File: PrintTreesInZigzag/PrintTreesInZigzag.cpp
/*
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) :
            val(x), left(NULL), right(NULL) {
    }
};
*/
class Solution {
public:
    vector<vector<int> > Print(TreeNode* root) {
        vector<vector<int>> res;
        if (root==NULL)
            return res;
        deque<TreeNode*> q;
        q.push_back(root);
        bool zigzag = true; // 从右打印为True，从左打印为False
        while (!q.empty())
        {
            int count = q.size();
            vector<int> out;
            TreeNode* node;
            while (count>0)
            {
                if (zigzag) //前取后放：从左向右，所以从前边取，后边放入
                {
                    node = q.front();
                    q.pop_front();
                    if (node->left)
                        q.push_back(node->left);
                    if (node->right)
                        q.push_back(node->right);
                } 
                else  // 后取前放：从右向左，从后边取，前边放入
                {
                    node = q.back();
                    q.pop_back();
                    if (node->right)
                        q.push_front(node->right);
                    if (node->left)
                        q.push_front(node->left);
                }
                out.push_back(node->val);
                count--;
            }
            res.push_back(out);
            zigzag = !zigzag;
        }
        return res;
    }
};

File: PrintTreesInZigzag/PrintTreesInZigzag.py
class Solution:
    def Print(self, root):
        # write code here
        res = []
        if root is None:
            return res
        import collections
        q = collections.deque()
        q.append(root)
        zigzag = True
        while q:
            n = len(q)
            out = []
            node = None
            while n>0:
                if zigzag:
                    node = q.popleft()
                    if node.left:
                        q.append(node.left)
                    if node.right:
                        q.append(node.right)
                else:
                    node = q.pop()
                    if node.right:
                        q.appendleft(node.right)
                    if node.left:
                        q.appendleft(node.left)
                out.append(node.val)
                n-=1
            res.append(out[:])
            zigzag = not zigzag
        return res
